File: scripts/extract_build_regulations.mjs
/**
 * Extracts the build-regulations workbook into public/data/build_regulations.json.
 *
 * Run from repo root:
 *   npm run extract:regulations
 *   node scripts/extract_build_regulations.mjs
 */
import { createHash } from 'node:crypto';
import { copyFileSync, existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import XLSX from 'xlsx';

const EXCEL_PATH_CANDIDATES = [
    'docs/data/regulations_data.xlsx',
    'public/data/regulations_data.xlsx',
];

const SUPPORTED_COUNTRIES = new Set(['Germany', 'Greece', 'Ireland', 'France']);
const BUILD_REGULATIONS_XLSX_HEADERS = [
    'Technology',
    'Country',
    'Nuts',
    'Nuts Name',
    'Variable',
    'Year Decision',
    'Year Ended',
    'Status',
    'Policy Effect',
    'Installation Type',
    'Installation Scale',
    'Location Or Characteristics',
    'Min Or Max',
    'Multiple Conditions',
    'Legally Binding',
    'Explicitly Mentioned',
    'Value 1',
    'Unit 1',
    'Condition 1',
    'Value 2',
    'Unit 2',
    'Condition 2',
    'Value 3',
    'Unit 3',
    'Condition 3',
    'Value 4',
    'Unit 4',
    'Condition 4',
    'Source Name',
    'Source Id',
    'Source Section',
    'Source Link',
    'Source Alternative',
    'Text Original',
    'Text Translation',
    'Miscellaneous',
    'Inactive Detail',
    'Notes Updated Laws',
    'Validated',
    'Record Type',
    'Serial Number',
    'Added In Version',
    'Status Changed In Version',
];

function findExcel() {
    const found = EXCEL_PATH_CANDIDATES.find((path) => existsSync(path));
    if (!found) {
        throw new Error('Could not find regulations_data.xlsx in expected locations.');
    }
    return found;
}

function cleanStr(value) {
    if (value == null) return null;
    if (typeof value === 'number' && Number.isNaN(value)) return null;
    const text = String(value).replace(/\r/g, '').trim();
    return text || null;
}

function cleanNum(value) {
    if (value == null) return null;
    if (typeof value === 'number' && !Number.isNaN(value)) return value;
    const match = String(value).match(/-?\d+(?:[.,]\d+)?/);
    if (!match) return null;
    const parsed = Number.parseFloat(match[0].replace(',', '.'));
    return Number.isNaN(parsed) ? null : parsed;
}

function isPlaceholder(value) {
    const text = cleanStr(value);
    if (!text) return true;
    return /^(n\/?a|na|null|none|-|\.{2,}\d*)$/i.test(text);
}

function capitalizeWords(value) {
    return value
        .trim()
        .toLowerCase()
        .replace(/\s+/g, ' ')
        .replace(/\b[a-z]/g, (char) => char.toUpperCase());
}

function normalizeCountry(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    const lower = raw.toLowerCase();
    if (lower.startsWith('germany')) return 'Germany';
    if (lower.startsWith('greece')) return 'Greece';
    if (lower.startsWith('ireland')) return 'Ireland';
    if (lower.startsWith('france')) return 'France';
    return capitalizeWords(raw);
}

function normalizeNuts(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    const nuts = raw.replace(/\s+/g, '').toUpperCase();
    return /^[A-Z]{2}[A-Z0-9]*$/.test(nuts) ? nuts : null;
}

function normalizeNutsName(value, nuts) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    if (raw.toUpperCase() === String(nuts || '').toUpperCase()) return null;
    return capitalizeWords(raw);
}

function normalizeYesNo(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    const lower = raw.toLowerCase();
    if (lower.startsWith('y')) return 'yes';
    if (lower.startsWith('n')) return 'no';
    return lower;
}

function normalizeStatus(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return raw.toLowerCase();
}

function normalizePolicyEffect(value) {
    const raw = cleanStr(value);
    const lower = (raw || 'constraining').toLowerCase();
    return lower.includes('promot') ? 'promoting' : 'constraining';
}

function normalizeVariable(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return raw.toLowerCase().replace(/\s+/g, ' ');
}

function normalizeInstallationType(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return raw.toLowerCase().replace(/\s+/g, '_');
}

function normalizeInstallationScale(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return capitalizeWords(raw.replace(/_/g, ' '));
}

function normalizeMinOrMax(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return raw.toLowerCase();
}

function normalizeMultilineText(value) {
    const raw = cleanStr(value);
    if (!raw || isPlaceholder(raw)) return null;
    return raw.replace(/\n+/g, '\n');
}

function buildColmap(columns) {
    const map = new Map();
    for (const column of columns) {
        if (column == null) continue;
        map.set(String(column).trim().toLowerCase(), column);
    }
    return map;
}

function getCell(row, colmap, ...names) {
    for (const name of names) {
        const key = String(name).trim().toLowerCase();
        const column = colmap.get(key);
        if (column != null) return row[column];
    }
    return null;
}

function compareRules(a, b) {
    return (
        String(a.country).localeCompare(String(b.country))
        || String(a.nuts).localeCompare(String(b.nuts))
        || String(a.kind).localeCompare(String(b.kind))
        || String(a.variable).localeCompare(String(b.variable))
        || Number(a.year_decision || 0) - Number(b.year_decision || 0)
        || String(a.serial_number || '').localeCompare(String(b.serial_number || ''))
    );
}

function compareWpa(a, b) {
    return (
        String(a.country).localeCompare(String(b.country))
        || String(a.nuts).localeCompare(String(b.nuts))
        || String(a.indicator || '').localeCompare(String(b.indicator || ''))
    );
}

function formatTechnologyLabel(value) {
    if (!value) return null;
    const normalized = String(value).trim().toLowerCase();
    if (normalized === 'wind') return 'Wind';
    if (normalized === 'solar') return 'Solar';
    if (normalized === 'ev') return 'Electric Vehicles';
    if (normalized === 'wind_priority_area') return 'Wind Priority Area';
    return capitalizeWords(String(value));
}

function formatVariableLabel(value) {
    if (!value) return null;
    return capitalizeWords(String(value).replace(/^\d+_/, '').replace(/_/g, ' '));
}

function flattenRuleForWorkbook(rule) {
    const row = {
        Technology: formatTechnologyLabel(rule.kind),
        Country: rule.country || null,
        Nuts: rule.nuts || null,
        'Nuts Name': rule.nuts_name || null,
        Variable: formatVariableLabel(rule.variable),
        'Year Decision': rule.year_decision ?? null,
        'Year Ended': rule.year_ended ?? null,
        Status: rule.status ? capitalizeWords(rule.status) : null,
        'Policy Effect': rule.policy_effect ? capitalizeWords(rule.policy_effect) : null,
        'Installation Type': rule.installation_type
            ? capitalizeWords(String(rule.installation_type).replace(/_/g, ' '))
            : null,
        'Installation Scale': rule.installation_scale || null,
        'Location Or Characteristics': rule.location_or_characteristics || null,
        'Min Or Max': rule.min_or_max ? capitalizeWords(rule.min_or_max) : null,
        'Multiple Conditions': rule.multiple_conditions || null,
        'Legally Binding': rule.legally_binding ? capitalizeWords(rule.legally_binding) : null,
        'Explicitly Mentioned': rule.explicitly_mentioned ? capitalizeWords(rule.explicitly_mentioned) : null,
        'Source Name': rule.source_name || null,
        'Source Id': rule.source_id || null,
        'Source Section': rule.source_section || null,
        'Source Link': rule.source_link || null,
        'Source Alternative': rule.source_alternative || null,
        'Text Original': rule.text_original || null,
        'Text Translation': rule.text_translation || null,
        Miscellaneous: rule.miscellaneous || null,
        'Inactive Detail': rule.inactive_detail ? capitalizeWords(rule.inactive_detail) : null,
        'Notes Updated Laws': rule.notes_updated_laws ? capitalizeWords(rule.notes_updated_laws) : null,
        Validated: rule.validated ? capitalizeWords(rule.validated) : null,
        'Record Type': 'Regulation',
        'Serial Number': rule.serial_number || null,
        'Added In Version': rule.added_in_version || 'V1.0',
        'Status Changed In Version': rule.status_changed_in_version || rule.added_in_version || 'V1.0',
    };

    (rule.values || []).slice(0, 4).forEach((entry, index) => {
        const slot = index + 1;
        row[`Value ${slot}`] = entry?.value ?? null;
        row[`Unit ${slot}`] = entry?.unit ?? null;
        row[`Condition ${slot}`] = entry?.condition ? capitalizeWords(entry.condition) : null;
    });

    return row;
}

function flattenWpaForWorkbook(area) {
    return {
        Technology: 'Wind Priority Area',
        Country: area.country || null,
        Nuts: area.nuts || null,
        'Nuts Name': area.nuts_name || null,
        Variable: formatVariableLabel(area.indicator),
        'Year Decision': null,
        'Year Ended': null,
        Status: area.status ? capitalizeWords(area.status) : 'Active',
        'Policy Effect': null,
        'Installation Type': null,
        'Installation Scale': null,
        'Location Or Characteristics': null,
        'Min Or Max': null,
        'Multiple Conditions': null,
        'Legally Binding': null,
        'Explicitly Mentioned': null,
        'Value 1': null,
        'Unit 1': null,
        'Condition 1': null,
        'Value 2': null,
        'Unit 2': null,
        'Condition 2': null,
        'Value 3': null,
        'Unit 3': null,
        'Condition 3': null,
        'Value 4': null,
        'Unit 4': null,
        'Condition 4': null,
        'Source Name': null,
        'Source Id': null,
        'Source Section': null,
        'Source Link': area.source_link || null,
        'Source Alternative': null,
        'Text Original': area.text_original || null,
        'Text Translation': area.text_translation || null,
        Miscellaneous: null,
        'Inactive Detail': null,
        'Notes Updated Laws': null,
        Validated: null,
        'Record Type': 'Wind Priority Area',
        'Serial Number': area.serial_number || null,
        'Added In Version': area.added_in_version || 'V1.0',
        'Status Changed In Version': area.status_changed_in_version || area.added_in_version || 'V1.0',
    };
}

function buildWorkbookRows(rules, wpa) {
    return [...rules.map(flattenRuleForWorkbook), ...wpa.map(flattenWpaForWorkbook)].sort(
        (a, b) =>
            String(a.Country || '').localeCompare(String(b.Country || ''))
            || String(a.Nuts || '').localeCompare(String(b.Nuts || ''))
            || String(a.Technology || '').localeCompare(String(b.Technology || ''))
            || String(a.Variable || '').localeCompare(String(b.Variable || ''))
            || Number(a['Year Decision'] || 0) - Number(b['Year Decision'] || 0)
    );
}

function parseSheet(rows, kind) {
    if (!rows.length) return [];
    const colmap = buildColmap(Object.keys(rows[0] || {}));
    const rules = [];

    rows.forEach((row, index) => {
        const country = normalizeCountry(getCell(row, colmap, 'Country'));
        const nuts = normalizeNuts(getCell(row, colmap, 'NUTS'));
        const variable = normalizeVariable(getCell(row, colmap, 'Variable', 'variable'));
        if (!country || !SUPPORTED_COUNTRIES.has(country) || !nuts || !variable) return;

        const statusRaw = normalizeStatus(getCell(row, colmap, 'status', 'active_inactive', 'active/inactive'));
        const addedInVersion = cleanStr(getCell(row, colmap, 'added_in_version')) || 'V1.0';
        const statusChangedInVersion = cleanStr(getCell(row, colmap, 'status_changed_in_version')) || addedInVersion;
        const serialNumber = cleanStr(getCell(row, colmap, 'serial_number', 'serial number'));
        const policyTypeRaw = cleanStr(getCell(row, colmap,
            'policy_type',
            'policy type',
            'type_of_policy',
            'constraining_promoting',
            'constraining_promoting ',
            'constraining/promoting',
            'type of policy',
        ));
        const policyEffect = normalizePolicyEffect(policyTypeRaw);

        const sourceId = cleanStr(getCell(row, colmap, 'Source_ID', 'source_id'));
        const sourceName = normalizeMultilineText(getCell(row, colmap, 'Source_name', 'source_name'));
        const yearDecision = cleanNum(getCell(row, colmap, 'Year_decision', 'year_decision'));
        const yearEndedRaw = cleanStr(getCell(row, colmap, 'Year_ended', 'year_ended'));
        const yearEnded = yearEndedRaw && !/^n\/?a$/i.test(yearEndedRaw)
            ? cleanNum(yearEndedRaw)
            : null;
        let policyId = serialNumber || sourceId;
        if (!policyId) {
            const payload = [
                kind,
                country,
                nuts,
                variable,
                String(yearDecision || ''),
                sourceName || '',
                cleanStr(getCell(row, colmap, 'Text_translation', 'text_translation')) || '',
            ].join('|');
            policyId = `gen_${createHash('sha1').update(payload).digest('hex').slice(0, 12)}`;
        }

        const rule = {
            kind,
            row_index: index + 2,
            serial_number: serialNumber,
            added_in_version: addedInVersion,
            status_changed_in_version: statusChangedInVersion,
            policy_id: policyId,
            policy_effect: policyEffect,
            policy_type_raw: policyTypeRaw,
            nuts,
            nuts_name: normalizeNutsName(getCell(row, colmap, 'NUTS_Name', 'NUTS_NAME', 'nuts_name'), nuts),
            country,
            year_decision: yearDecision == null ? null : Math.trunc(yearDecision),
            year_ended: yearEnded == null ? null : Math.trunc(yearEnded),
            location_or_characteristics: capitalizeWords(
                cleanStr(getCell(row, colmap, 'Location_or_characteristics', 'location_or_characteristics')) || ''
            ) || null,
            variable,
            installation_type: normalizeInstallationType(getCell(row, colmap, 'Installation_type', 'installation_type')),
            installation_scale: normalizeInstallationScale(getCell(row, colmap, 'Installation_scale', 'installation_scale')),
            min_or_max: normalizeMinOrMax(getCell(row, colmap, 'Minimum_or_maximum', 'min_or_max')),
            multiple_conditions: capitalizeWords(
                cleanStr(getCell(row, colmap, 'Multiple_conditions_attribute', 'multiple_conditions_attribute')) || ''
            ) || null,
            values: [],
            legally_binding: normalizeYesNo(getCell(row, colmap, 'Legally_binding', 'legally_binding')),
            explicitly_mentioned: normalizeYesNo(
                getCell(row, colmap, 'WT_explicitly_mentioned')
                || getCell(row, colmap, 'Solar_explicitly_mentioned')
                || getCell(row, colmap, 'EV_explicitly_mentioned')
            ),
            source_name: sourceName,
            source_id: sourceId,
            source_section: cleanStr(getCell(row, colmap, 'Source_section', 'source_section')),
            source_link: cleanStr(getCell(row, colmap, 'Source_link', 'source_link')),
            source_alternative: cleanStr(getCell(row, colmap, 'Source_alternative', 'source_alternative')),
            text_original: normalizeMultilineText(getCell(row, colmap, 'Text_original', 'text_original')),
            text_translation: normalizeMultilineText(getCell(row, colmap, 'Text_translation', 'text_translation')),
            miscellaneous: normalizeMultilineText(getCell(row, colmap, 'Miscellaneous', 'miscellaneous')),
            status: statusRaw,
            active: statusRaw,
            inactive_detail: cleanStr(getCell(row, colmap, 'inactive_policy_status', 'inactive_detail', 'inactive_reason')),
            supersedes_policy: cleanStr(getCell(row, colmap, 'supersedes_policy', 'Supersedes_policy')),
            overwritten_by_row: cleanNum(getCell(row, colmap,
                'overwritten_by_row',
                'overwritten_policy_replacement',
                'replaced_by_row',
                'replacing_policy_row',
            )),
            notes_updated_laws: cleanStr(getCell(row, colmap, 'notes_updated_laws', 'notes - updated laws')),
            validated: normalizeYesNo(getCell(row, colmap, 'Validated_by_experts', 'validated_by_experts')),
        };

        for (let i = 1; i <= 4; i += 1) {
            const value = cleanNum(getCell(row, colmap, `Value_${i}`, `value_${i}`));
            const unit = cleanStr(getCell(row, colmap, `Unit_${i}`, `unit_${i}`));
            const condition = cleanStr(getCell(row, colmap, `Condition_${i}`, `condition_${i}`));
            if (value != null || unit || condition) {
                rule.values.push({ value, unit, condition });
            }
        }

        rules.push(rule);
    });

    return rules.sort(compareRules);
}

function parseWpa(rows) {
    if (!rows.length) return [];
    const colmap = buildColmap(Object.keys(rows[0] || {}));
    const out = [];

    rows.forEach((row, index) => {
        const nuts = normalizeNuts(getCell(row, colmap, 'NUTS'));
        const country = normalizeCountry(getCell(row, colmap, 'COUNTRY', 'Country'));
        const indicator = cleanStr(getCell(row, colmap, 'INDICATOR', 'indicator'));
        const sourceLink = cleanStr(getCell(row, colmap, 'SOURCE_LINK', 'source_link'));
        const textOriginal = normalizeMultilineText(getCell(row, colmap, 'TEXT_ORIGINAL', 'text_original'));
        const textTranslation = normalizeMultilineText(getCell(row, colmap, 'TEXT_TRANSLATION', 'text_translation'));

        if (!nuts || !country || !SUPPORTED_COUNTRIES.has(country) || !indicator) return;
        if (isPlaceholder(sourceLink) && isPlaceholder(textOriginal) && isPlaceholder(textTranslation)) return;

        const addedInVersion = cleanStr(getCell(row, colmap, 'added_in_version')) || 'V1.0';
        const statusChangedInVersion = cleanStr(getCell(row, colmap, 'status_changed_in_version')) || addedInVersion;

        out.push({
            kind: 'wind_priority_area',
            row_index: index + 2,
            serial_number: cleanStr(getCell(row, colmap, 'serial_number', 'serial number')),
            added_in_version: addedInVersion,
            status_changed_in_version: statusChangedInVersion,
            nuts,
            nuts_name: normalizeNutsName(getCell(row, colmap, 'NUTS_NAME', 'NUTS_Name', 'nuts_name'), nuts),
            country,
            indicator,
            source_link: isPlaceholder(sourceLink) ? null : sourceLink,
            text_original: textOriginal,
            text_translation: textTranslation,
            status: 'active',
            active: 'active',
        });
    });

    return out.sort(compareWpa);
}

function summarize(rules) {
    const countries = {};
    const byVariable = {};
    const byYear = {};
    for (const rule of rules) {
        countries[rule.country] = (countries[rule.country] || 0) + 1;
        byVariable[rule.variable] = (byVariable[rule.variable] || 0) + 1;
        if (rule.year_decision) {
            byYear[rule.year_decision] = (byYear[rule.year_decision] || 0) + 1;
        }
    }
    return {
        by_country: Object.fromEntries(Object.entries(countries).sort(([a], [b]) => a.localeCompare(b))),
        by_variable: Object.fromEntries(Object.entries(byVariable).sort(([a], [b]) => a.localeCompare(b))),
        by_year: Object.fromEntries(Object.entries(byYear).sort(([a], [b]) => Number(a) - Number(b))),
    };
}

function main() {
    const path = findExcel();
    const workbook = XLSX.readFile(path);
    const sheetRows = (name) => XLSX.utils.sheet_to_json(workbook.Sheets[name] || {}, { defval: null });

    const wind = parseSheet(sheetRows('Wind regulations'), 'wind');
    const solar = parseSheet(sheetRows('Solar regulations'), 'solar');
    const ev = parseSheet(sheetRows('EV charging regulations'), 'ev');
    const wpa = parseWpa(sheetRows('WPA_GR'));
    const rules = [...wind, ...solar, ...ev].sort(compareRules);
    const summary = summarize(rules);

    const output = {
        meta: {
            source: path,
            generated_at: new Date().toISOString(),
            rule_count: rules.length,
            wind_priority_area_count: wpa.length,
            ...summary,
            sheets: ['Wind regulations', 'Solar regulations', 'EV charging regulations', 'WPA_GR'],
        },
        rules,
        wind_priority_areas: wpa,
    };

    const publicPath = join('public', 'data', 'build_regulations.json');
    const docsPath = join('docs', 'data', 'build_regulations.json');
    const publicWorkbookPath = join('public', 'data', 'build_regulations.xlsx');
    const docsWorkbookPath = join('docs', 'data', 'build_regulations.xlsx');
    mkdirSync(join('public', 'data'), { recursive: true });
    mkdirSync(join('docs', 'data'), { recursive: true });
    const payload = `${JSON.stringify(output, null, 2)}\n`;
    writeFileSync(publicPath, payload, 'utf8');
    copyFileSync(publicPath, docsPath);
    const workbookRows = buildWorkbookRows(rules, wpa);
    const exportWorkbook = XLSX.utils.book_new();
    const sheet = XLSX.utils.json_to_sheet(workbookRows, { header: BUILD_REGULATIONS_XLSX_HEADERS });
    XLSX.utils.book_append_sheet(exportWorkbook, sheet, 'Building_Regulations');
    XLSX.writeFile(exportWorkbook, publicWorkbookPath, { compression: true });
    copyFileSync(publicWorkbookPath, docsWorkbookPath);
    console.log(`Wrote ${publicPath}: ${rules.length} rules, ${wpa.length} wind-priority-area rows.`);
    console.log(`Wrote ${docsWorkbookPath}: ${workbookRows.length} export rows.`);
    console.log('Countries:', summary.by_country);
}

main();

File: scripts/extract_build_regulations.py
"""
Extracts the build-regulations workbook into public/data/build_regulations.json.

Preferred entry point (Node, no openpyxl version issues):
    npm run extract:regulations

Python alternative:
    python scripts/extract_build_regulations.py
"""
from __future__ import annotations
import json
import math
import os
import re
import sys
import hashlib
from collections import defaultdict

import pandas as pd

EXCEL_PATH_CANDIDATES = [
    "docs/data/regulations_data.xlsx",
    "public/data/regulations_data.xlsx",
    "docs/data/regulations_data.xls",
    "public/data/regulations_data.xls",
    "../regulations_data.xlsx",
    "regulations_data.xlsx",
    "../regulations_data.xls",
    "regulations_data.xls",
    "../D2.2.1.1_Data colllection_regulations for energy infrastructure_feb25.xlsx",
    "D2.2.1.1_Data colllection_regulations for energy infrastructure_feb25.xlsx",
    os.path.expanduser(
        "~/Desktop/docs/regulations/D2.2.1.1_Data colllection_regulations for energy infrastructure_feb25.xlsx"
    ),
]


def find_excel() -> str:
    for path in EXCEL_PATH_CANDIDATES:
        if os.path.exists(path):
            return path
    sys.exit("Could not find the build-regulations workbook in expected locations.")


def clean_str(v):
    if v is None:
        return None
    if isinstance(v, float) and math.isnan(v):
        return None
    s = str(v).strip()
    return s or None


def clean_num(v):
    if v is None:
        return None
    if isinstance(v, (int, float)) and not math.isnan(float(v)):
        return float(v)
    if isinstance(v, str):
        m = re.search(r"-?\d+(?:[.,]\d+)?", v)
        if m:
            try:
                return float(m.group(0).replace(",", "."))
            except ValueError:
                return None
    return None


def normalize_country(c):
    s = clean_str(c)
    if not s:
        return None
    s = s.strip()
    if s.lower().startswith("germany"):
        return "Germany"
    if s.lower().startswith("greece"):
        return "Greece"
    if s.lower().startswith("ireland"):
        return "Ireland"
    if s.lower().startswith("france"):
        return "France"
    return s


def normalize_nuts(n):
    s = clean_str(n)
    if not s:
        return None
    return re.sub(r"\s+", "", s).upper()


def parse_sheet(df: pd.DataFrame, kind: str):
    colmap = {str(c).strip().lower(): c for c in df.columns}

    def col(*names: str):
        for name in names:
            key = str(name).strip().lower()
            if key in colmap:
                return colmap[key]
        return None

    def get(row, *names: str):
        c = col(*names)
        return row.get(c) if c is not None else None

    rules = []
    for idx, row in df.iterrows():
        country = normalize_country(get(row, "Country"))
        if not country:
            continue
        nuts = normalize_nuts(get(row, "NUTS"))
        if not nuts:
            continue
        status_raw = clean_str(get(row, "status", "active_inactive", "active/inactive", "Active_inactive"))
        added_in_version = clean_str(get(row, "added_in_version")) or "V1.0"
        status_changed_in_version = clean_str(get(row, "status_changed_in_version")) or added_in_version
        serial_number = clean_str(get(row, "serial_number", "serial number"))
        status = status_raw.lower() if status_raw else None
        policy_type_raw = clean_str(get(
            row,
            "policy_type",
            "policy type",
            "type_of_policy",
            "constraining_promoting",
            "constraining_promoting ",
            "constraining/promoting",
            "type of policy",
        ))
        policy_type = (policy_type_raw or "constraining").lower()
        if "promot" in policy_type:
            policy_effect = "promoting"
        else:
            policy_effect = "constraining"

        source_id = clean_str(get(row, "Source_ID", "source_id"))
        source_name = clean_str(get(row, "Source_name", "source_name"))
        variable = clean_str(get(row, "Variable", "variable"))
        year_decision = int(clean_num(get(row, "Year_decision", "year_decision")) or 0) or None
        year_ended_raw = clean_str(get(row, "Year_ended", "year_ended"))
        year_ended = None
        if year_ended_raw and not re.match(r"^n/?a$", year_ended_raw, re.I):
            year_ended_val = clean_num(year_ended_raw)
            year_ended = int(year_ended_val) if year_ended_val is not None else None
        policy_id = serial_number or source_id
        if not policy_id:
            id_payload = "|".join([
                kind,
                country or "",
                nuts or "",
                variable or "",
                str(year_decision or ""),
                source_name or "",
                clean_str(get(row, "Text_translation", "text_translation")) or "",
            ])
            policy_id = f"gen_{hashlib.sha1(id_payload.encode('utf-8')).hexdigest()[:12]}"

        rule = {
            "kind": kind,
            "row_index": int(idx) + 2,  # +2 for spreadsheet row index with header
            "serial_number": serial_number,
            "added_in_version": added_in_version,
            "status_changed_in_version": status_changed_in_version,
            "policy_id": policy_id,
            "policy_effect": policy_effect,
            "policy_type_raw": policy_type_raw,
            "nuts": nuts,
            "nuts_name": clean_str(get(row, "NUTS_Name", "NUTS_NAME", "nuts_name")),
            "country": country,
            "year_decision": year_decision,
            "year_ended": year_ended,
            "location_or_characteristics": clean_str(get(row, "Location_or_characteristics", "location_or_characteristics")),
            "variable": variable,
            "installation_type": clean_str(get(row, "Installation_type", "installation_type")),
            "installation_scale": clean_str(get(row, "Installation_scale", "installation_scale")),
            "min_or_max": clean_str(get(row, "Minimum_or_maximum", "min_or_max")),
            "multiple_conditions": clean_str(get(row, "Multiple_conditions_attribute", "multiple_conditions_attribute")),
            "values": [],
            "legally_binding": clean_str(get(row, "Legally_binding", "legally_binding")),
            "explicitly_mentioned": clean_str(
                get(row, "WT_explicitly_mentioned")
                or get(row, "Solar_explicitly_mentioned")
                or get(row, "EV_explicitly_mentioned")
            ),
            "source_name": source_name,
            "source_id": source_id,
            "source_section": clean_str(get(row, "Source_section", "source_section")),
            "source_link": clean_str(get(row, "Source_link", "source_link")),
            "source_alternative": clean_str(get(row, "Source_alternative", "source_alternative")),
            "text_original": clean_str(get(row, "Text_original", "text_original")),
            "text_translation": clean_str(get(row, "Text_translation", "text_translation")),
            "miscellaneous": clean_str(get(row, "Miscellaneous", "miscellaneous")),
            "status": status_raw,
            "active": status_raw,  # backward compatibility for existing clients
            "inactive_detail": clean_str(get(row, "inactive_policy_status", "inactive_detail", "inactive_reason")),
            "supersedes_policy": clean_str(get(row, "supersedes_policy", "Supersedes_policy")),
            "overwritten_by_row": clean_num(get(
                row,
                "overwritten_by_row",
                "overwritten_policy_replacement",
                "replaced_by_row",
                "replacing_policy_row",
            )),
            "notes_updated_laws": clean_str(get(row, "notes_updated_laws", "notes - updated laws")),
            "validated": clean_str(get(row, "Validated_by_experts", "validated_by_experts")),
        }
        for i in (1, 2, 3, 4):
            v = clean_num(get(row, f"Value_{i}", f"value_{i}"))
            u = clean_str(get(row, f"Unit_{i}", f"unit_{i}"))
            c = clean_str(get(row, f"Condition_{i}", f"condition_{i}"))
            if v is not None or u or c:
                rule["values"].append({"value": v, "unit": u, "condition": c})
        if not rule["variable"]:
            continue
        rules.append(rule)
    return rules


def parse_wpa(df: pd.DataFrame):
    out = []
    for idx, row in df.iterrows():
        nuts = normalize_nuts(row.get("NUTS"))
        country = normalize_country(row.get("COUNTRY"))
        indicator = clean_str(row.get("INDICATOR"))
        if not nuts or not country or not indicator:
            continue
        source_link = clean_str(row.get("SOURCE_LINK"))
        text_original = clean_str(row.get("TEXT_ORIGINAL"))
        text_translation = clean_str(row.get("TEXT_TRANSLATION"))
        if source_link and source_link.upper() in {"N/A", "NA"} and not text_original and not text_translation:
            continue
        added_in_version = clean_str(row.get("added_in_version")) or "V1.0"
        status_changed_in_version = clean_str(row.get("status_changed_in_version")) or added_in_version
        out.append({
            "kind": "wind_priority_area",
            "row_index": int(idx) + 2,
            "serial_number": clean_str(row.get("serial_number")),
            "added_in_version": added_in_version,
            "status_changed_in_version": status_changed_in_version,
            "nuts": nuts,
            "nuts_name": clean_str(row.get("NUTS_NAME")),
            "country": country,
            "indicator": indicator,
            "source_link": source_link,
            "text_original": text_original,
            "text_translation": text_translation,
            "status": "active",
            "active": "active",
        })
    return out


def main():
    path = find_excel()
    sheets = pd.read_excel(path, sheet_name=None, header=0)

    wind = parse_sheet(sheets.get("Wind regulations", pd.DataFrame()), "wind")
    solar = parse_sheet(sheets.get("Solar regulations", pd.DataFrame()), "solar")
    ev = parse_sheet(sheets.get("EV charging regulations", pd.DataFrame()), "ev")
    wpa = parse_wpa(sheets.get("WPA_GR", pd.DataFrame()))

    rules = wind + solar + ev

    countries = defaultdict(int)
    by_variable = defaultdict(int)
    by_year = defaultdict(int)
    for r in rules:
        countries[r["country"]] += 1
        by_variable[r["variable"]] += 1
        if r["year_decision"]:
            by_year[r["year_decision"]] += 1

    output = {
        "meta": {
            "source": "D2.2.1.1_Data collection_regulations for energy infrastructure_feb25.xlsx",
            "rule_count": len(rules),
            "by_country": dict(sorted(countries.items())),
            "by_variable": dict(sorted(by_variable.items())),
            "by_year": dict(sorted(by_year.items())),
            "sheets": ["Wind regulations", "Solar regulations", "EV charging regulations", "WPA_GR"],
        },
        "rules": rules,
        "wind_priority_areas": wpa,
    }

    out_path = os.path.join("public", "data", "build_regulations.json")
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f"Wrote {out_path}: {len(rules)} rules, {len(wpa)} wind-priority-area rows.")


if __name__ == "__main__":
    main()
